File: 463. Island Perimeter/Python/Solution.py
class Solution(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """

        def is_water(i, j):
            return i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]) or grid[i][j] == 0

        directions = ((0, -1), (0, 1), (-1, 0), (1, 0))
        perimeter = 0
        for (i, row) in enumerate(grid):
            for (j, item) in enumerate(row):
                if item == 1:
                    for direction in directions:
                        if is_water(i + direction[0], j + direction[1]):
                            perimeter += 1
        return perimeter

File: 463. Island Perimeter/C++/Solution.cpp
class Solution {
public:
    int islandPerimeter(vector <vector<int>> &grid) {
        auto isWater = [&grid](int i, int j) {
            return i < 0 || j < 0 || i >= grid.size() || j >= grid[0].size() || grid[i][j] == 0;
        };
        vector <vector<int>> directions{{0,  -1}, {0,  1}, {-1, 0}, {1,  0}};
        int perimeter = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == 1) {
                    for (auto &direction : directions) {
                        if (isWater(i + direction[0], j + direction[1])) {
                            perimeter++;
                        }
                    }
                }
            }
        }
        return perimeter;
    }
};